pkg/ipamd/ipamd.go

```diff
@@ -2042,10 +2042,14 @@ func (c *IPAMContext) isDatastorePoolTooLow() bool {
 		totalIPs = maxIpsPerPrefix
 	}
 
-	poolTooLow := available < totalIPs*warmTarget || (warmTarget == 0 && available == 0)
+	poolTooLow := c.dataStore.GetENIs() == 0 || available < totalIPs*warmTarget
 	if poolTooLow {
 		log.Debugf("IP pool is too low: available (%d) < ENI target (%d) * addrsPerENI (%d)", available, warmTarget, totalIPs)
 		c.logPoolStats(stats)
+	} else if warmTarget == 0 && c.dataStore.IsIPStarvedInDS() {
+		poolTooLow = available <= 0
```

Comment:
We appear to be setting `poolTooLow` in the `else if` check of `if poolTooLow`? Why can't we enhance the original statement in #2045? Also, we're setting ds.ipStarved back to false before the reconciler adds additional IPs/ENIs. Shouldn't we do it after the ENI or IP add succeeds and datastore has a non zero count of available addresses?

Reply:
poolTooLow doesn't check the special case of warm eni is zero. We want to avoid reconciler adding new ENI to DS if WARM_ENI is 0 and no extra IP is required. When poolTwooLOw is set in `else if`, it just overrides the variable's value for return. 
I prefer to reset the DS ipStarved back to `false` as soon as we can to avoid possible race from multiple requests. If IP allocation fails, the kubelet will try again and DS will set the property to true which will be recognized by ipam reconciler again.
Please let me know if this could cause failure or significant latency. I am also ok to keep the `true` state longer since the worst case would be ipam attaching extra ENI and will remove it eventually.

Reply:
I agree with Apurup to merge the poolTooLow into one logic. Also instead of a flag, can't we just walk thru the ENIs and see if IPs are starved?

Reply:
Watching the flag can avoid walking thru the ENIs every reconciling cycle. Moreover the checking is not necessary in most of use cases. When we use flag to deliver the signal, only the DS encounter IP starving, ipamd takes action.

Reply:
But the ENI's max will be just few so shouldn't be a big walk. Maybe lets sync up offline?

Reply:
Yeah, we can discuss offline. I think that the key is we need to do ad-hoc on the signal of IP starving instead of watching on ENIs since watching on ENIs without a signal can lead to the original ENIs churning problem.

Reply:
@jayanthvn @haouc Thanks for doing this PR! Where there any take aways from the offline discussion? Can I help on this issue in anyway?